backend/src/Edna/Upload/Service.hs

```diff
@@ -103,18 +102,21 @@ uploadFile' projSqlId@(SqlId proj) methodSqlId@(SqlId method)
       expFileId <- UQ.insertExperimentFile projId methodId (fcMetadata fc)
         description fileName fileBytes
 
-      forM_ (toPairs fileMeasurements) $
-        \(targetName, TargetMeasurements targetMeasurements) ->
-          forM_ (toPairs targetMeasurements) $ \(compoundName, measurements) -> do
-            let
-              getId :: HasCallStack => Text -> HashMap Text (SqlId a) -> SqlId a
-              getId name =
-                fromMaybe (error $ "unexpected name: " <> name) .
-                view (at name)
-            let targetId = getId targetName targetToId
-            let compoundId = getId compoundName compoundToId
-            insertExperiment expFileId compoundId targetId measurements
-
+      experiments <- sortBy
+        (\a b -> (a ^. _1 . _2, a ^. _1 . _1) `compare` (b ^. _1 . _2, b ^. _1 . _1)) <$>
```

Comment:
I see a lot of sorting in various places, can you explain why we need it?

Reply:
Because `returning` functionality from sql doesn't guarantee order in which you insert/update values and we need it. There is `with` functionality with which we can sort our `returning`, but beam doesn't support it (it support only `with select`, not `with insert`)

Reply:
Backend

Reply:
Ok, where?

Reply:
I can probably figure it out myself, but it'll be simpler if you explain.